doc: add board terminology section to board porting guide (#69802)
zephyrproject-rtos/zephyr: doc/hardware/porting/board_porting.rst

.. _example-application conversion Pull Request: https://github.com/zephyrproject-rtos/example-application/pull/58
.. _conversion script: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/utils/board_v1_to_v2.py

.. _hw_support_hierarchy:

Hardware support hierarchy
**************************

Zephyr's hardware support hierarchy has the following levels, from most to least

> specific:

Reviewer: Not sure I agree that the list goes from "most to least specific", but it's not a blocker.

- :term:`board`, which has one or more
- :term:`SoC`, each of which optionally belong to a
- :term:`SoC series`, which in turn may optionally belong to a
- :term:`SoC family`, which has one or more

> - :term:`CPU Cluster`, each containing one or more

Author: minor nit, this could give the impression that it is the SoC family which defines / has one or more CPU cluster.

In fact it is the SoC which defines the CPU cluster, not the family.

Reviewer 2: Well spotted.

   | :ref:`nrf5340dk <nrf5340dk_nrf5340>`       | nrf5340/cpuapp           | nRF5340     | nRF53              | Nordic nRF         | Arm Cortex-M33 | Arm Cortex-M         |
   |                                            +--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
   |                                            | nrf5340/cpunet           | nRF5340     | nRF53              | Nordic nRF         | Arm Cortex-M33 | Arm Cortex-M         |
   +--------------------------------------------+--------------------------+-------------+--------------------+--------------------+----------------+----------------------+
   | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8ml8/a53             | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-A53 | Arm64                |
   |                                            +--------------------------+-------------+--------------------+--------------------+----------------+----------------------+

>    |                                            | mimx8ml8/m7              | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-M7  | Arm Cortex-M         |

Reviewer 3: architecture isn't really right, M4 would be armv7, M33 would be armv8

Reviewer 2: yes, good point. I will fix that.

>    |                                            | mimx8ml8/m7              | i.MX8M Plus | i.MX8M             | NXP i.MX           | Arm Cortex-M7  | Arm Cortex-M         |

Reviewer: "Arm Cortex-M" isn't really the architecture in Zephyr terminology, though - is it?

Reviewer 2: Yep, see the comment above

classifies and implements hardware support.
This section focuses on the terminology used around hardware support, and in
particular when defining and working with boards and SoCs.

The overall set of terms used around the concept of board in Zephyr is depicted
in the image below, which uses the :ref:`bl5340_dvk` board as reference.

.. figure:: board/board-terminology.png
   :width: 500px
   :align: center
   :alt: Board terminology diagram

   Board terminology diagram

The diagram shows the different terms that are used to describe boards:

- The :term:`board name`: ``bl5340_dvk``
- The optional :term:`board revision`: ``1.2.0``
- The :term:`board qualifiers`, that optionally describe the :term:`SoC`,
  :term:`CPU cluster` and :term:`variant`: ``nrf5340/cpuapp/ns``
- The :term:`board target`, which uniquely identifies a combination of the above
  and can be used to specify the hardware to build for when using the tooling
  provided by Zephyr: ``bl5340_dvk@1.2.0/nrf5340/cpuapp/ns``

Formally this can also be seen as:

.. code-block:: console

   {board name}[@{revision}][/{board qualifiers}]

which can be extended to:

.. code-block:: console

>    {board name}[@{revision}][/{SoC}][/{CPU cluster}][/{variant}]

Author: to me this indicates that you can have a CPU cluster or variant without a SoC.

Should this have been:
```suggestion
   {board name}[@{revision}][/{SoC}[/{CPU cluster}][/{variant}]]
```

Reviewer 2: Indeed

>    {board name}[@{revision}][/{SoC}[/{CPU cluster}][/{variant}]]

Reviewer 4: The combination of `[ ]`s and `{ }`s makes this really hard to parse visually IMO
Please consider the following:

```
Formally this can also be seen as
:samp:`{board name}[@{revision}][/{board qualifiers}]`, which can be extended to
:samp:`{board name}[@{revision}][/{SoC}[/{CPU cluster}][/{variant}]]`.
```

<img width="854" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/ff9ffcb5-36d2-40cc-8a73-e488b454b7cc">

If a board contains only one single-core SoC, then the SoC can be omitted from

> the board target. This implies that, if the board does not define any board

Reviewer 4: ```suggestion
the board target. This implies that if the board does not define any board
```

.. _board_and_identifiers:

> Board names, identifiers and targets

Reviewer 5: qualifiers?

Reviewer 2: Sorry, this is not ready for review yet. Working on it now

   +--------------------------------------------+-----------------------+-------------+---------------+---------------+----------------+--------------+
   | :ref:`rv32m1_vega <rv32m1_vega>`           | openisa_rv32m1/ri5cy  | RV32M1      | (Not used)    | (Not used)    | RI5CY          | RISC-V       |
   +--------------------------------------------+-----------------------+-------------+---------------+---------------+----------------+--------------+
   | :ref:`nrf5340dk <nrf5340dk_nrf5340>`       | nrf5340/cpuapp        | nRF5340     | nRF53         | Nordic nRF    | Arm Cortex-M33 | Arm          |
   |                                            +-----------------------+-------------+---------------+---------------+----------------+--------------+
   |                                            | nrf5340/cpunet        | nRF5340     | nRF53         | Nordic nRF    | Arm Cortex-M33 | Arm          |
   +--------------------------------------------+-----------------------+-------------+---------------+---------------+----------------+--------------+

>    | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8m/a53            | i.MX8M Plus | i.MXM8M A53   | NXP i.MX      | Arm Cortex-A53 | Arm64        |

Reviewer 6: s/i.MXM8M A53/i.MX8M A53

Author: see https://github.com/zephyrproject-rtos/zephyr/pull/69802#discussion_r1514519675

>    | :ref:`mimx8mp_evk <imx8mp_evk>`            | mimx8m/a53            | i.MX8M Plus | i.MX8M        | NXP i.MX      | Arm Cortex-A53 | Arm64        |

Reviewer 6: Is `mimx8m` the correct identifier?
In boards/ we have `mimx8ml8`.
See:
A53 -  https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.yaml#L7
M7 - https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_m7.yaml#L7

Reviewer 2: Done

   |                                            +-----------------------+-------------+---------------+---------------+----------------+--------------+

>    |                                            | mimx8m/m7             | i.MX8M Plus | i.MXM8MM M4   | NXP i.MX      | Arm Cortex-M7  | Arm          |

Reviewer 6: The SoC Series is wrong here - it should be i.MX8M M7 - https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/imx/soc.yml#L32

Reviewer 6: Also, I would add here another entry for other arch, like Xtensa:
HiFi4 | Xtensa |  mimx8m/adsp | i.MX8M Plus | i.MX8M ADSP | NXP i.MX | HIFi4

   Board extensions are not allowed in any module referenced in Zephyr's
   ``west.yml`` manifest file. Any board changes are required to be submitted to
   the main Zephyr repository.

.. _board_terminology:

> Terminology

Reviewer 5: would be nicer if you used some formatting for this type of thing, so it is easy to read and find:

https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html#create-a-glossary

Reviewer 5: would be nice as a followup to have few diagrams showing all of this and the layering and how things relate and what type of "setups" are possible. a picture is worth 1000 words....

Author: > would be nicer if you used some formatting for this type of thing, so it is easy to read and find:
> 
> https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html#create-a-glossary

Thanks for the link, will take a look at updating.
I just followed what we do here: https://docs.zephyrproject.org/latest/project/index.html#terminology

Perhaps long term we should try to have a uniform terminology style, fyi @kartben

Reviewer 5: now that you mention this, you should remove the "/" from identifier :)

Reviewer 5: and yes, we can start with a link to this table.

Author: > > would be nicer if you used some formatting for this type of thing, so it is easy to read and find:
> > https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html#create-a-glossary
> 
> Thanks for the link, will take a look at updating.

Done

Author: > and yes, we can start with a link to this table.

added a reference to the section as I couldn't get referencing the table directly to work.

Author: > now that you mention this, you should remove the "/" from identifier :)

done

***********

> - board: a physical piece of hardware or an emulated board.

Reviewer 2: Please add an example for each of the terms

   Board extensions are not allowed in any module referenced in Zephyr's
   ``west.yml`` manifest file. Any board changes are required to be submitted to
   the main Zephyr repository.

.. _board_terminology:

> Terminology

Reviewer 4: sorry if I missed something but this list is almost exactly a copy-paste of the "Boards, SoCs, etc." section, so I am not sure there is a reason for the two to still co-exist?

Author: there was a discussion on having a terminology section, see https://discord.com/channels/720317445772017664/1194671118422913114/1214257604205744158 and before comments. A proposal to which I agree, especially as it also makes it easier to reference.

But you're right that `Boards, SoCs, etc.` contains part of same, but the reason I didn't remove it from there is because I still think it makes sense to have a brief description at that location where we introduces the concepts.

At the same time, a reader will expect that a terminology section describes all important terms, as a form of quick reference. So hard to argue those term should not be there.

Reviewer 2: I think we can live with this duplication. Longer term what we need is a linkable global Glossary instead, like we do in NCS:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/glossary.html

Reviewer 4: > I think we can live with this duplication. Longer term what we need is a linkable global Glossary instead, like we do in NCS:
> https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/glossary.html
> 

Exactly :) What's wrong with using the glossary we already have? The text that's used in "Boards, SoCs, ..." can stay there as a legend for the diagram, but should also link to proper entries in the glossary (and this will be a good opportunity to update existing entries such as "SoC")
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/glossary.rst

***********

.. glossary::

    board
      a physical piece of hardware or an emulated board.

    board identifier
      a name which identifies the exact SoC, CPU cluster, or board
      variant for the current build. The board identifier is separated from the

>       board name by a ``/``. Similar the SoC is also separated from the CPU

Reviewer 7: ```suggestion
      board name by a ``/``. Similarly, the SoC is also separated from the CPU
```

      cluster and variant by a ``/``.

    board revision
      a specific revision of a board when the board exists in multiple revisions
      and where a given revision needs adjustments to devicetree or Kconfig
      files for correct behavior.

>      SoC

Author: indentation fixed, and list sorted alphabetically.

>      SoC

Author: > the rendered page looks very funny:

Even though I have forced pushed with alphabetically sorted list and indentation fixed, the doc build output has not been updated.

>      SoC

Reviewer 2: This is fixed now as far as I can tell:
https://builds.zephyrproject.io/zephyr/pr/69802/docs/hardware/porting/board_porting.html#terminology

       the exact system on a chip the board's CPU is part of

     SoC series
       a smaller group of tightly related SoCs

     SoC family
       a wider group of SoCs with similar characteristics

     CPU Cluster
       a cluster of one or more CPU cores.

     CPU core
       a particular CPU in an architecture

     Architecture
       an instruction set architecture

Examples:

The board :ref:`nrf5340dk_nrf5340` is a single SoC board where the SoC defines
two CPU clusters, ``cpuapp`` and ``cpunet``. One of the CPU clusters, the
``cpuapp``, furthermore defines a non-secure board variant, ``ns``.

The board with identifier like ``nrf5340dk/nrf5340/cpuapp`` can be read as:

- ``nrf5340dk``: This part is before the first ``/`` and is the board name.
- ``nrf5340/cpuapp``: is the board identifier, and can be further divided into

  - ``nrf5340``: The SoC, which is ``nrf5340``

>   - ``cpuapp``: The CPU cluster ``cpuapp`` which for this SoC is a single CPU
>     core of Cortex-m33. The number of cores in a CPU cluster cannot be

Reviewer 7: ```suggestion
  - ``cpuapp``: The CPU cluster ``cpuapp``, which is a single Cortex-M33 CPU
    core for this SoC. The number of cores in a CPU cluster cannot be
```

    determined from the board identifier.

The ``cpuapp`` also defines a non-secure board variant, which is identified as
``nrf5340dk/nrf5340/cpuapp/ns``. The first part of the board identifier has been
explained above, and the extra part is read as:

- ``ns``: a board variant,, in this case ``ns`` is a common variant name is
  Zephyr denoting a non-secure build for boards supporting :ref:`tfm`.

> Not all SoCs defines CPU clusters or board variants. For example a simple board
> like the :ref:`thingy52_nrf52832` contains a single SoC with no CPU cluster and

Reviewer 7: ```suggestion
Not all SoCs define CPU clusters or board variants. For example a simple board
like the :ref:`thingy52_nrf52832` contains a single SoC with no CPU clusters and
```

no board variants.
For ``thingy52`` the board with identifier like ``thingy52/nrf52832`` can be
read as:

- ``thingy52``: This part is before the first ``/`` and is the board name.

> - ``nrf52832``: is the board identifier, and as there is no more levels then
>   it's in this case identical the SoC, which is ``nrf52832``.

Reviewer 7: ```suggestion
- ``nrf52832``: is the board identifier, and as there are no more levels then
  it's in this case identical to the SoC, which is ``nrf52832``.
```
